### PR TITLE
[Merged by Bors] - refactor: move GaussSum directly under NumberTheory

### PR DESCRIPTION
**Pull request by MichaelStollBayreuth**
This moves `NumberTheory.LegendreSymbol.GaussSum` to `NumberTheory.GaussSum`.
Rationale: Gauss sums are useful beyond quadratic reciprocity (which was the original motivation to add them to Mathlib).

We use the opportunity to deal with some porting notes.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by MichaelStollBayreuth** [comment]
!bench

**Comment by MichaelStollBayreuth** [comment]
!bench

**Comment by MichaelStollBayreuth** [comment]
```text
General information:
  build:                                               -15.814 * 10⁹ (-0.0133%)

No file got slower by at least 10⁹ instructions.

No file got slower by at least 10%.

Files that got faster by at least 10⁹ instructions:
  Mathlib.Analysis.NormedSpace.PiTensorProduct.InjectiveSeminorm:
                                                       -2.5181 * 10⁹ (-0.915%)
  Mathlib.RingTheory.Localization.Basic:               -1.3839 * 10⁹ (-1.25%)

No file got faster by at leat 10%.
```

File: Mathlib/NumberTheory/GaussSum.lean [hunk 1]
```diff
@@ -158,11 +158,10 @@ theorem gaussSum_frob (χ : MulChar R R') (ψ : AddChar R R') :
 /-- For a quadratic character `χ` and when the characteristic `p` of the target ring
 is a unit in the source ring, the `p`th power of the Gauss sum of`χ` and `ψ` is
 `χ p` times the original Gauss sum. -/
--- Porting note: Added `nonrec` to avoid error `failed to prove termination`
-nonrec theorem MulChar.IsQuadratic.gaussSum_frob (hp : IsUnit (p : R)) {χ : MulChar R R'}
+theorem MulChar.IsQuadratic.gaussSum_frob (hp : IsUnit (p : R)) {χ : MulChar R R'}
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/GaussSum.lean:161]: This is what `nonrec` is for, so I'm not sure we should avoid it.

**Review by Ruben-VandeVelde — APPROVED** [review]

File: Mathlib/NumberTheory/GaussSum.lean (hunk 1 above)
Comment (MichaelStollBayreuth) [reply Mathlib/NumberTheory/GaussSum.lean:161]: I was thinking of [this Zulip message](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/.E2.80.9Cdeprecated.E2.80.9D.20error.20message/near/438479016) here, but I have no strong opinion.

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**